zipline-loader/src/main/kotlin/app/cash/zipline/loader/ZiplineLoader.kt

```diff
@@ -0,0 +1,156 @@
+package app.cash.zipline.loader
+
+import app.cash.zipline.Zipline
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.async
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.Semaphore
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.sync.withPermit
+import okio.ByteString
+
+/**
+ * Gets code from an HTTP server or a local cache,
+ * and loads it into a zipline instance. This attempts
+ * to load code as quickly as possible, and will
+ * concurrently download and load code.
+ */
+class ZiplineLoader(
+// TODO add caching
+//  val cacheDirectory: Path,
+//  val cacheMaxSizeInBytes: Int = 100 * 1024 * 1024,
+  val client: ZiplineHttpClient,
+) {
+  /*
+  file a.js:
+  globalThis.loadedFiles = globalThis.loadedFiles || [];
+  globalThis.loadedFiles += 'A';
+   */
+
+  /**
+   * This returns once all of the modules in [manifest] have been loaded into [zipline].
+   * This function can be canceled, but doing so leaves [zipline] in an undefined state.
+   *
+   * version zero:
+   * download each file sequentially
+   * call Zipline.loadJsModule()
+   *
+   * version one:
+   * download each file sequentially if not stored on disk
+   * store files on disk using okio.FileSystem
+   * gotcha: atomicMove()
+   * gotcha: download fails, this will leave garbage around forever
+   * call Zipline.loadJsModule()
+   *
+   * version two:
+   * same as version one, but it doesn't leave garbage around forever
+   * probably Sqlite to store downloads in flight
+   * call Zipline.loadJsModule()
+   *
+   * version three:
+   * download in parallel
+   * load as soon as downloads finish
+   * honor topo sort
+   * https://github.com/square/okhttp/blob/master/okhttp/src/main/kotlin/okhttp3/internal/cache/DiskLruCache.kt
+   *
+   * version four:
+   * prune old files
+   * note that a file that has been pruned might also be loaded shortly after pruning; perhaps
+   * even between deleting on disk and deleting in DB
+   */
+  suspend fun load(scope: CoroutineScope, zipline: Zipline, manifest: ZiplineManifest) {
+    val ziplineMutex = Mutex()
```

Comment:
I was thinking about this, and I think instead of the mutex we aught to `async` the work on the zipline’s dispatcher, which is guaranteed single-threaded.